### PR TITLE
tests: network: Prefer internal IPs first

### PR DESCRIPTION
**Pull request by r4f4**
#### What type of PR is this?

/kind cleanup

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by aojea** [comment]
This test was this way for a long time, is this a real issue ?

**Comment by r4f4** [comment]
> This test was this way for a long time, is this a real issue ?

We have faced this in our CI when deploying clusters with public-only subnets. We figured either pods are required to be able to connect to external IPs, in which case the test is correct as is, and our failure is a legitimate test failure that we need to address; or else pods are not required to be able to connect to external IPs, in which case the test is possibly wrong and needs to be fixed to use the internal IP instead.

**Comment by aojea** [comment]
These tests grew organically and most probably the people that added them is no longer active in the project so is difficult to have historical context. I've done some archeology and I could find that at the beginning this test only used external addresses

https://github.com/kubernetes/kubernetes/commit/81ff1f87f0b79e70b876d798ee0de18d9861a11b

In addition, the tests are supposed to test the project itself,  adding firewalls or rules on top a cluster is modifying the scenario under test and is not something we do during development ... however, I understand that people start to use test as a way of verifying their environments, but that is not the purpose of the e2e tests

I think that your reasoning is legit, but in the other side, if we prefer InternalIP, ExternalIP are not going to be tested anymore (InternalIP are always present) , so we'll regress in coverage in a certain way ...

**Comment by danwinship** [comment]
> I've done some archeology and I could find that at the beginning this test only used external addresses

No, originally nodes (or rather, "minions") only had a single IP, and after NodeAddresses and the InternalIP/ExternalIP distinction was added, there was a long period where the original single IP was preserved as the "LegacyHostIP" and most of the tests used that. But that was deprecated and eventually went away (#44830) and all the tests had to be updated.

So in the beginning, the test used the single node IP, and then when LegacyHostIP went away, it got changed to use ExternalIP, except that that caused problems, so then it was changed to use either ExternalIP or InternalIP.

> In addition, the tests are supposed to test the project itself, adding firewalls or rules on top a cluster is modifying the scenario under test and is not something we do during development

The initial description was kind of vague, but the issue here is that in some clouds, node-to-node traffic via ExternalIPs will be treated as cloud-to-internet traffic rather than cloud-internal traffic for billing purposes. Thus, a CI run where tests use ExternalIPs costs more than a CI run where pods use InternalIPs!

> if we prefer InternalIP, ExternalIP are not going to be tested anymore (InternalIP are always present), so we'll regress in coverage in a certain way ...

Yes... an ExternalIP "is, in some way, intended to be more usable from outside the cluster then an internal IP, though no specific semantics are defined". In particular, we can't assume ExternalIPs are globally reachable like LoadBalancer IPs would be.

Maybe the best approach would be to use InternalIPs for the majority of tests, but have a handful of explicit ExternalIP tests, if there are ExternalIPs. (IMO we should also make sure that no conformance tests assume that pods can reach External IPs.)

**Comment by spiffxp** [comment]
/ok-to-test
I think this is a reasonable thing to be re-examined but I'm not sure it's wise to introduce this late in the release cycle

**Comment by aojea** [comment]
> /ok-to-test I think this is a reasonable thing to be re-examined but I'm not sure it's wise to introduce this late in the release cycle

@spiffxp this topic rang a bell and I couldn't find it today, we have already discussed this, see

https://github.com/kubernetes/kubernetes/issues/90764#issuecomment-790767411 

In summary, we should not require ExternalIPs for Conformance, but we should keep testing ExternalIPs in sig network.

We may take the same solution we took there https://github.com/kubernetes/kubernetes/pull/98791

 ... but agree that now is late in the cycle to do this, this code is part of a lot of e2e tests in conformance

File: test/e2e/framework/network/utils.go [hunk 1]
```diff
@@ -817,15 +817,15 @@ func (config *NetworkingTestConfig) setup(selector map[string]string) {
 		family = v1.IPv6Protocol
 		secondaryFamily = v1.IPv4Protocol
 	}
-	// Get Node IPs from the cluster, ExternalIPs take precedence
-	config.NodeIP = e2enode.FirstAddressByTypeAndFamily(nodeList, v1.NodeExternalIP, family)
```
Comment (aojea) [inline test/e2e/framework/network/utils.go:821]: Make this parametrizable, I've added functional options some time ago, you can do something lile
```
// PreferExternalAddresses prefer node External Addresses for the tests
func PreferExternalAddresses(config *NetworkingTestConfig) {
	config.PreferExternalAddresses = true
}
```
```

Comment (aojea) [reply test/e2e/framework/network/utils.go:821]: and here you just use the bool to pick the address

```

if config.PreferExternalAddress {
  config.NodeIP = e2enode.FirstAddressByTypeAndFamily(nodeList, v1. NodeExternalIP, family)
}
if config.NodeIP == "" {
		config.NodeIP = e2enode.FirstAddressByTypeAndFamily(nodeList, v1.NodeInternalIP, family)
}
```

or something like this

**Comment by aojea** [comment]
> > /ok-to-test I think this is a reasonable thing to be re-examined but I'm not sure it's wise to introduce this late in the release cycle
> 
> @spiffxp this topic rang a bell and I couldn't find it today, we have already discussed this, see
> 
> [#90764 (comment)](https://github.com/kubernetes/kubernetes/issues/90764#issuecomment-790767411)
> 
> In summary, we should not require ExternalIPs for Conformance, but we should keep testing ExternalIPs in sig network.
> 
> We may take the same solution we took there #98791
> 
> ... but agree that now is late in the cycle to do this, this code is part of a lot of e2e tests in conformance

/lgtm

 this is the same solution adopted for the similar situation described in my previous comment 

https://github.com/kubernetes/kubernetes/pull/113799#issuecomment-1311619565

/area conformance

defer approval and milestone to @spiffxp @johnbelamaric

File: test/e2e/framework/network/utils.go [hunk 2]
```diff
@@ -205,6 +210,8 @@ type NetworkingTestConfig struct {
 	// The kubernetes namespace within which all resources for this
 	// config are created
 	Namespace string
+	// Whether to prefer node External Addresses for the tests
+	PreferExternalAddresses bool
```
Comment (spiffxp) [inline test/e2e/framework/network/utils.go:214]: I think @aojea's pattern suggestion is great.  However as written this still changes the default behavior for everything to prefer internal addresses by default.  

I would suggest we either not merge this until post-1.26, or you flip this around so it defaults to the existing behavior of preferring external addresses (and you can selectively change for certain tests until test freeze).

**Review by spiffxp — CHANGES_REQUESTED** [review]

File: test/e2e/framework/network/utils.go (hunk 2 above)
Comment (r4f4) [reply test/e2e/framework/network/utils.go:216]: I don't have any strong preferences to either but I've changed the code to just parameterize the option for now without change of behavior.

**Comment by aojea** [comment]
I'd prefer to avoid 2 PRs and do just one iteration on this:
1. adding the option and switching the default at the same time
2. adding the option now and switching the default later

**Comment by r4f4** [comment]
> I'd prefer to avoid 2 PRs and do just one iteration on this:
> 
>     1. adding the option and switching the default at the same time

I changed back to this option so we can wait to merge post 1.26.

/hold

**Comment by aojea** [comment]
/milestone v1.27

seems there is consensus for next release, I'll go ahead and approve, unhold once master reopens

/approve

**Comment by jluhrsen** [comment]
> /milestone v1.27
> 
> seems there is consensus for next release, I'll go ahead and approve, unhold once master reopens
> 
> /approve

@aojea, can we try to get this in now? 1.26 is out for a week or so, so I'm assuming 1.27(master) is opened now right?

**Comment by aojea** [comment]
> > /milestone v1.27
> > seems there is consensus for next release, I'll go ahead and approve, unhold once master reopens
> > /approve
> 
> @aojea, can we try to get this in now? 1.26 is out for a week or so, so I'm assuming 1.27(master) is opened now right?

yeah, absolutely, sorry, it slipped

/retest
/hold cancel
/lgtm
/approve

**Comment by jluhrsen** [comment]
> > > /milestone v1.27
> > > seems there is consensus for next release, I'll go ahead and approve, unhold once master reopens
> > > /approve
> > 
> > 
> > @aojea, can we try to get this in now? 1.26 is out for a week or so, so I'm assuming 1.27(master) is opened now right?
> 
> yeah, absolutely, sorry, it slipped
> 
> /retest /hold cancel /lgtm /approve

:)

that was fast. thank you!

**Pull request merged**